src/KOKKOS/metatensor_system_kokkos.cpp

```diff
@@ -0,0 +1,412 @@
+    auto expanded_arange = torch::arange(max_number_of_neighbors, torch::TensorOptions().dtype(torch::kInt32).device(device)).unsqueeze(0).expand({total_n_atoms, -1});
+    auto neighbor_2d_mask = expanded_arange < numneigh_torch.unsqueeze(1);
+
+    auto expanded_arange_other_dim = torch::arange(total_n_atoms, torch::TensorOptions().dtype(torch::kInt32).device(device)).unsqueeze(1).expand({-1, max_number_of_neighbors});
+    auto index_for_ilist = expanded_arange_other_dim.masked_select(neighbor_2d_mask);
+    auto centers_tensor = ilist_torch.index_select(0, index_for_ilist);
+
+    auto neighbors_tensor = neighbors_torch.masked_select(neighbor_2d_mask);
+
+    // change centers and neighbors to the original atom ids
+    auto centers_tensor_original_id = original_atom_id_tensor.index_select(0, centers_tensor);
+    auto neighbors_tensor_original_id = original_atom_id_tensor.index_select(0, neighbors_tensor);
+
+    // create torch tensor with the positions
+    auto positions_tensor = torch::from_blob(
+        positions_kokkos.data(),
+        {total_n_atoms, 3},
+        torch::TensorOptions().dtype(torch::kFloat64).device(device)
+    ).to(dtype);
+
+    // The following code is a direct translation of the code in the non-Kokkos version (MetaTensorSystemAdaptor::setup_neighbors_remap),
+    // but rewritten in torch to use the GPU
+    for (auto& cache: caches_) {
+        // half list mask, if necessary (TODO: change names! This could modify the tensors outside the loop if more than one NL!)
+        auto full_list = cache.options->full_list();
+
+        torch::Tensor centers_tensor_original_id_full_or_half;
+        torch::Tensor neighbors_tensor_original_id_full_or_half;
+        torch::Tensor centers_tensor_full_or_half;
+        torch::Tensor neighbors_tensor_full_or_half;
+        if (full_list) {
+            centers_tensor_full_or_half = centers_tensor;
+            neighbors_tensor_full_or_half = neighbors_tensor;
+            centers_tensor_original_id_full_or_half = centers_tensor_original_id;
+            neighbors_tensor_original_id_full_or_half = neighbors_tensor_original_id;
+        } else {
+            auto half_list_mask = centers_tensor_original_id <= neighbors_tensor_original_id;
+            centers_tensor_full_or_half = centers_tensor.masked_select(half_list_mask);
+            neighbors_tensor_full_or_half = neighbors_tensor.masked_select(half_list_mask);
+            centers_tensor_original_id_full_or_half = centers_tensor_original_id.masked_select(half_list_mask);
+            neighbors_tensor_original_id_full_or_half = neighbors_tensor_original_id.masked_select(half_list_mask);
+        }
+
+        // distance mask
+        auto interatomic_vectors = positions_tensor.index_select(0, neighbors_tensor_full_or_half) - positions_tensor.index_select(0, centers_tensor_full_or_half);
+        auto distance_mask = torch::sum(interatomic_vectors.pow(2), 1) < cache.cutoff*cache.cutoff;
```

Comment:
`torch.vector_norm` might be a bit faster here

Reply:
I leave this up to you